pkg/queue/breaker.go

```diff
@@ -62,69 +64,77 @@ func NewBreaker(params BreakerParams) *Breaker {
 	}
 	sem := newSemaphore(params.MaxConcurrency, params.InitialCapacity)
 	return &Breaker{
-		pendingRequests: make(chan struct{}, params.QueueDepth+params.MaxConcurrency),
-		sem:             sem,
+		totalSlots: int64(params.QueueDepth + params.MaxConcurrency),
+		sem:        sem,
 	}
 }
 
+// tryAcquirePending tries to acquire a slot on the pending "queue".
+func (b *Breaker) tryAcquirePending() bool {
+	for {
+		cur := atomic.LoadInt64(&b.inFlight)
+		if cur == b.totalSlots {
+			return false
+		}
+		if atomic.CompareAndSwapInt64(&b.inFlight, cur, cur+1) {
```

Comment:
I convinced myself I understood out what's going on here, but it took me a couple reads :). Probably worth a comment.